### PR TITLE
New data source: azurerm_mysql_server

### PR DESCRIPTION
**Pull request by Lucretius**
Resolves https://github.com/terraform-providers/terraform-provider-azurerm/issues/4924

This is not a resource I have ever used, but I followed along the existing resource to build out the data source, and reused the tests (adding in the data source at the end of each) and checking the values.

### CODE REVIEW

File: website/docs/d/mysql_server.html.markdown [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+* `ssl_minimal_tls_version_enforced` - The minimum TLS version to support for this MySQL Server.
+
+* `storage_mb` -  Max storage allowed for this MySQL Server.
+
+* `threat_detection_policy` - Threat detection policy configuration, known in the API as Server Security Alerts Policy. The `threat_detection_policy` block supports fields documented below.
+
+* `tags` - A mapping of tags to assign to the resource.
+---
+
+A `identity` block supports the following:
```
Comment (manicminer) [inline website/docs/d/mysql_server.html.markdown:72]:
```suggestion
An `identity` block exports the following:
```

File: website/docs/d/mysql_server.html.markdown [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+   
+* `type` -  The Type of Identity used for this MySQL Server.
+
+* `principal_id` - The Client ID of the Service Principal assigned to this MySQL Server.
+
+* `tenant_id` - The ID of the Tenant the Service Principal is assigned in.
+---
+
+a `threat_detection_policy` block supports the following:
```
Comment (manicminer) [inline website/docs/d/mysql_server.html.markdown:81]:
```suggestion
A `threat_detection_policy` block exports the following:
```

File: website/docs/d/mysql_server.html.markdown [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+* `public_network_access_enabled` - Whether or not public network access is allowed for this MySQL Server.
+
+* `ssl_enforcement_enabled` -  Specifies if SSL should be enforced on connections for this MySQL Server.
+
+* `ssl_minimal_tls_version_enforced` - The minimum TLS version to support for this MySQL Server.
+
+* `storage_mb` -  Max storage allowed for this MySQL Server.
+
+* `threat_detection_policy` - Threat detection policy configuration, known in the API as Server Security Alerts Policy. The `threat_detection_policy` block supports fields documented below.
```
Comment (manicminer) [inline website/docs/d/mysql_server.html.markdown:67]:
```suggestion
* `threat_detection_policy` - Threat detection policy configuration, known in the API as Server Security Alerts Policy. The `threat_detection_policy` block exports fields documented below.
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @Lucretius, thanks for this PR! I have some minor suggestions below, including a way to dry out the tests a bit. If we can fix these up, this looks good to merge.

**Comment by Lucretius** [comment]
Thanks @manicminer - I've accepted all the suggestions, thanks for that (wish Github let me accept them all in one commit instead of one at a time).  Also gives me a good idea of what to watch out for in the future.

**Comment by manicminer** [comment]
@Lucretius great! could you update the other tests too? (`testAccDataSourceAzureRMMySQLServer_basicWithIdentity`, `testAccDataSourceAzureRMMySQLServer_autogrow` and `testAccDataSourceAzureRMMySQLServer_complete`)

File: website/docs/d/mysql_server.html.markdown [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+output "id" {
+  value = data.azurerm_mysql_server.example.id
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the MySQL Server. Changing this forces a new resource to be created. This needs to be globally unique within Azure.
```
Comment (manicminer) [inline website/docs/d/mysql_server.html.markdown:31]:
```suggestion
* `name` - (Required) Specifies the name of the MySQL Server.
```

File: website/docs/d/mysql_server.html.markdown [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+* `resource_group_name` - (Required) The name of the resource group in which to create the MySQL Server. Changing this forces a new resource to be created.
```
Comment (manicminer) [inline website/docs/d/mysql_server.html.markdown:33]:
```suggestion
* `resource_group_name` - (Required) The name of the resource group for the MySQL Server.
```

File: azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+func testAccDataSourceAzureRMMySQLServer_basic(data acceptance.TestData, version string) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_mysql_server" "test" {
+  name                = "acctestmysqlsvr-%d"
+  resource_group_name = azurerm_resource_group.test.name
+}
+`, testAccAzureRMMySQLServer_basic(data, version), data.RandomInteger)
```
Comment (manicminer) [inline azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go:171]: The data source needs to reference an original resource attribute so Terraform can infer the correct order, as without it, it tries to look up the MySQL server before it's created.

```suggestion
  name                = azurerm_mysql_server.test.name
  resource_group_name = azurerm_resource_group.test.name
}
`, testAccAzureRMMySQLServer_basic(data, version))
```

File: azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+func testAccDataSourceAzureRMMySQLServer_basicWithIdentity(data acceptance.TestData, version string) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_mysql_server" "test" {
+  name                = "acctestmysqlsvr-%d"
+  resource_group_name = azurerm_resource_group.test.name
+}
+`, testAccAzureRMMySQLServer_basicWithIdentity(data, version), data.RandomInteger)
```
Comment (manicminer) [inline azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go:182]:
```suggestion
  name                = azurerm_mysql_server.test.name
  resource_group_name = azurerm_resource_group.test.name
}
`, testAccAzureRMMySQLServer_basic(data, version))
```

File: azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+func testAccDataSourceAzureRMMySQLServer_autogrow(data acceptance.TestData, version string) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_mysql_server" "test" {
+  name                = "acctestmysqlsvr-%d"
+  resource_group_name = azurerm_resource_group.test.name
+}
+`, testAccAzureRMMySQLServer_autogrow(data, version), data.RandomInteger)
```
Comment (manicminer) [inline azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go:193]:
```suggestion
  name                = azurerm_mysql_server.test.name
  resource_group_name = azurerm_resource_group.test.name
}
`, testAccAzureRMMySQLServer_basic(data, version))
```

File: azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+func testAccDataSourceAzureRMMySQLServer_complete(data acceptance.TestData, version string) string {
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_mysql_server" "test" {
+  name                = "acctestmysqlsvr-%d"
+  resource_group_name = azurerm_resource_group.test.name
+}
+`, testAccAzureRMMySQLServer_complete(data, version), data.RandomInteger)
```
Comment (manicminer) [inline azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go:204]:
```suggestion
  name                = azurerm_mysql_server.test.name
  resource_group_name = azurerm_resource_group.test.name
}
`, testAccAzureRMMySQLServer_basic(data, version))
```

File: azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+		CheckDestroy: testCheckAzureRMMySQLServerDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAzureRMMySQLServer_autogrow(data, mysqlVersion),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMMySQLServerExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "sku_name", "GP_Gen5_2"),
+					resource.TestCheckResourceAttr(data.ResourceName, "administrator_login", "acctestun"),
+					resource.TestCheckResourceAttr(data.ResourceName, "auto_grow_enabled", "true"),
+					resource.TestCheckResourceAttr(data.ResourceName, "ssl_minimal_tls_version_enforced", "TLS1_1"),
```
Comment (manicminer) [inline azurerm/internal/services/mysql/tests/mysql_server_data_source_test.go:126]: This isn't set in the config
```suggestion
```

**Review by manicminer — CHANGES_REQUESTED** [review]
@Lucretius Thanks for making those changes. When running the tests, I spotted some other issues that I missed earlier.

**Comment by Lucretius** [comment]
Thanks again for clarifying the issues with the tests, sorry there were so many!  I'll keep them in mind in future PRs.

**Comment by manicminer** [comment]
@Lucretius I pushed some additional minor test fixes (some of the configs were pointing at the wrong test config funcs and some attribute check values didn't match the configuration). The tests now pass. Thanks again!

Screenshot 2020-10-13 at 20 15 09

**Pull request merged**